common/nexus/incoming_service_registry.go

```diff
@@ -0,0 +1,355 @@
+	// Wait for service data to be initialized before long polling.
+	if _, err := r.serviceDataReady.Get(ctx); err != nil {
+		return err
+	}
+
+	r.dataLock.RLock()
+	currentTableVersion := r.tableVersion
+	r.dataLock.RUnlock()
+
+	resp, err := r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+		NextPageToken:         nil,
+		PageSize:              int32(r.config.refreshPageSize()),
+		LastKnownTableVersion: currentTableVersion,
+		Wait:                  true,
+	})
+	if err != nil {
+		return err
+	}
+
+	if resp.TableVersion == currentTableVersion {
+		// Long poll returned with no changes
+		return nil
+	}
+
+	// changes were returned by long poll so acquire write lock to prevent reads while updating
+	r.dataLock.Lock()
+	defer r.dataLock.Unlock()
+
+	if r.tableVersion >= resp.TableVersion {
+		// service data was updated while waiting for the write lock
+		return nil
+	}
+
+	currentTableVersion = resp.TableVersion
+	services := make(map[string]*nexus.IncomingService)
+	for _, service := range resp.Services {
+		services[service.Id] = service
+	}
+
+	currentPageToken := resp.NextPageToken
+	for currentPageToken != nil {
+		resp, err = r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+			NextPageToken:         currentPageToken,
+			PageSize:              int32(r.config.refreshPageSize()),
+			LastKnownTableVersion: currentTableVersion,
+			Wait:                  false,
+		})
+
+		if err != nil {
+			if errors.Is(err, p.ErrNexusTableVersionConflict) {
+				// Indicates table was updated during paging, so reset and start from the beginning
+				currentTableVersion, services, err = r.getAllServicesMatching(ctx)
+				if err != nil {
+					return err
+				}
+				break
+			}
+			return err
+		}
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		for _, service := range resp.Services {
+			services[service.Id] = service
+		}
+
+		currentPageToken = resp.NextPageToken
+	}
+
+	r.tableVersion = currentTableVersion
+	r.services = services
+	return nil
+}
+
+// initializeServices initializes the in-memory view of service data.
+// It first tries to load from matching service and falls back to querying persistence directly if matching is unavailable.
+func (r *IncomingServiceRegistry) initializeServices(ctx context.Context) error {
+	r.dataLock.Lock()
+	defer r.dataLock.Unlock()
+
+	if r.serviceDataReady.Ready() {
+		// Indicates service data was loaded by another thread while waiting for write lock.
+		return nil
+	}
+
+	tableVersion, services, err := r.getAllServicesMatching(ctx)
+	if err != nil {
+		// Fallback to persistence on matching error during initial load.
+		tableVersion, services, err = r.getAllServicesPersistence(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	r.tableVersion = tableVersion
+	r.services = services
+	if !r.serviceDataReady.Ready() {
+		r.serviceDataReady.Set(struct{}{}, nil)
+	}
```

Comment:
If you get here and the future is ready, it's a bug.

Reply:
That's true but I figured it was better not to panic if we did somehow get here and it's ready since I don't think it should be a fatal bug. I will add some extra logging though.